modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -493,15 +493,97 @@ def join(self, right, **kwargs):
+                if not isinstance(on, list):
+                    on = [on]
+                if len(on) != len(right.index.names):
+                    raise ValueError(
+                        'len(left_on) must equal the number of levels in the index of "right"'
+                    )
+                # `on` can include index name and column name in the same time,
+                # but they shouldn't be equally
+                intersection = pandas.Index(self.index.names).intersection(self.columns)
+                if not intersection.empty:
+                    intersected = list(filter(lambda label: label in intersection, on))
+                    raise ValueError(
+                        f"'{intersected[0]}' is both an index level and a column label, which is ambiguous."
+                    )
 
-            new_self = self.__constructor__(
-                self._modin_frame.apply_full_axis(1, map_func)
-            )
+                if how == "left":
+                    labels = _create_index(self, on, self.index, self.columns)
+                    right = right.reindex(
+                        axis=0, labels=labels, _reset_index=self.index
+                    )
+                    new_self = self.concat(1, right, join="left")
+                elif how == "inner":
+                    labels = _create_index(self, on, self.index, self.columns)
+                    inner_labels = labels.intersection(right.index)
+                    new_self = self
+                    if len(inner_labels) != len(labels):
+                        _, indexer = labels.reindex(inner_labels)
+                        if indexer is not None:
+                            new_index = new_self.index.take(indexer)
+                        else:
+                            new_index = new_self.index
+                        new_self = new_self.reindex(
+                            axis=0, labels=inner_labels, _reset_index=new_index
+                        )
+                    # we need always to reset index, but we may try to do it separately
+                    right = right.reindex(
+                        axis=0, labels=inner_labels, _reset_index=new_self.index
+                    )
+                    new_self = new_self.concat(1, right, join="left")
+                elif how == "right":
+                    labels = _create_index(self, on, self.index, self.columns)
+                    _, indexer = labels.reindex(right.index)
+                    if indexer is not None:
+                        # `take` doesn't work with indexer, where `-1` value can be
+                        new_index = self.index.take(indexer)
+                    else:
+                        new_index = self.index
+                    new_self = self.reindex(axis=0, labels=new_index)
+                    right.index = new_self.index
+                    new_self = new_self.concat(1, right, join="left")
+            else:
+                right = right.to_pandas()
```

Comment:
is there any way to avoid this to_pandas?

Reply:
I suppose that yes, but there is out of scope of the PR.